cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/DefaultAuthenticationContextBuilder.java

```diff
@@ -0,0 +1,177 @@
+    @Override
+    public boolean collect(final Authentication authentication) throws AuthenticationException {
+        if (this.authentications.add(authentication)) {
+            LOGGER.debug("Collected authentication event. Associated principal with this authentication is [{}]",
+                    authentication.getPrincipal());
+            return true;
+        }
+        LOGGER.warn("Failed to collect duplicate authentication event as it already exists.");
+        return false;
+    }
+
+    @Override
+    public AuthenticationContext build() {
+        final Authentication authentication = buildAuthentication();
+        if (authentication == null) {
+            throw new RuntimeException(new GeneralSecurityException("Authentication context cannot be produced"));
+        }
+        return new DefaultAuthenticationContext(authentication);
+    }
+
+    @Override
+    public void clear() {
+        this.authentications.clear();
+    }
+
+    private Authentication buildAuthentication() {
+        if (isEmpty()) {
+            LOGGER.warn("No authentication event has been recorded; CAS cannot finalize the authentication context");
+            return null;
+        }
+
+        final Map<String, Object> authenticationAttributes = new HashMap<>();
+        final Map<String, Object> principalAttributes = new HashMap<>();
+        final AuthenticationBuilder authenticationBuilder = DefaultAuthenticationBuilder.newInstance();
+
+        buildAuthenticationHistory(authenticationAttributes, principalAttributes, authenticationBuilder);
+
+        final Principal primaryPrincipal = getPrimaryPrincipal(principalAttributes);
+        authenticationBuilder.setPrincipal(primaryPrincipal);
+        LOGGER.debug("Determined primary authentication principal to be [{}]", primaryPrincipal);
+
+        authenticationBuilder.setAttributes(authenticationAttributes);
+        LOGGER.debug("Collected authentication attributes for this context are [{}]", authenticationAttributes);
+
+        final DateTime dt = DateTime.now();
+        authenticationBuilder.setAuthenticationDate(dt);
+        LOGGER.debug("Authentication context commenced at [{}]", dt);
+
+        return authenticationBuilder.build();
+
+    }
+
+    private void buildAuthenticationHistory(final Map<String, Object> authenticationAttributes,
+                                            final Map<String, Object> principalAttributes,
+                                            final AuthenticationBuilder authenticationBuilder) {
+
+        LOGGER.debug("Collecting authentication history based on [{}] authentication events", this.authentications.size());
+        for (final Authentication authn : this.authentications) {
+            final Principal authenticatedPrincipal = authn.getPrincipal();
+            LOGGER.debug("Evaluating authentication principal [{}] for inclusion in context", authenticatedPrincipal);
+
+            principalAttributes.putAll(authenticatedPrincipal.getAttributes());
```

Comment:
At first, I was a bit surprised by the merge of the principal attributes (and only that point), but thinking a little more about that, that's not surprising as multiple authentications occur, producing multiple principals and the identity is the sum of these principals.